session.c

```diff
@@ -467,88 +551,144 @@ int do_exec_windows(struct ssh *ssh, Session *s, const char *command, int pty) {
 	fcntl(pipein[1], F_SETFD, FD_CLOEXEC);
 	fcntl(pipeout[0], F_SETFD, FD_CLOEXEC);
 	fcntl(pipeerr[0], F_SETFD, FD_CLOEXEC);
+	
+	/* setup Environment varibles */ 
+	do {
+		static int environment_set = 0;
+
+		if (environment_set)
+			break;
 
-	/* setup Environment varibles */
-	setup_session_vars(s);
+		if (setup_session_vars(s) != 0)
+			goto cleanup;
+
+		environment_set = 1;
+	} while (0);
 
 	if (!in_chroot)
 		chdir(s->pw->pw_dir);
 
-	/* prepare exec - path used with CreateProcess() */
-	if (s->is_subsystem || (command && memcmp(command, "scp", 3) == 0)) {
-		/* relative or absolute */
-		if (command == NULL || command[0] == '\0')
-			fatal("expecting command for a subsystem");
-
-		if (command[1] == ':') /* absolute */
-			exec_command = xstrdup(command);
-		else {/*relative*/
-			const int command_len = strlen(progdir) + 1 + strlen(command) + (strlen(sftp_exe) - strlen(INTERNAL_SFTP_NAME));
-			exec_command = malloc(command_len);
-			if (exec_command == NULL)
-				fatal("%s, out of memory", __func__);
-						
-			cmd = exec_command;
-			memcpy(cmd, progdir, strlen(progdir));
-			cmd += strlen(progdir);
-			*cmd++ = '\\';
-
-			/* In windows, INTERNAL_SFTP is supported via sftp-server.exe.
-			 * This is a deviation from the UNIX implementation that hosts sftp-server within sshd.
-			 * If sftp-server were to be hosted within sshd for Windows, following would be needed
-			 *  - Impersonate client user
-			 *  - call sftp-server-main
-			 *
-			 * SSHD service account would need impersonate privilege to impersonate client user, 
-			 * thereby needing elevation of SSHD account privileges
-			 * Apart from slight performance gain (by hosting sftp in process), there isn't a clear 
-			 * gain with this option over using and spawning sftp-server.exe.
-			 * Hence going with the later option. 
-			 */
-			if(IS_INTERNAL_SFTP(command)) {
-				memcpy(cmd, sftp_exe, strlen(sftp_exe) + 1);
-				cmd += strlen(sftp_exe);
-				
-				// copy the arguments (if any).
-				if(strlen(command) > strlen(INTERNAL_SFTP_NAME)) {
-					argp = (char*)command + strlen(INTERNAL_SFTP_NAME);
-					memcpy(cmd, argp, strlen(argp)+1);
-				}
+#define CMDLINE_APPEND(P, S)		\
+do {					\
+	int _S_len = strlen(S);		\
+	memcpy((P), (S), _S_len);		\
+	(P) += _S_len;			\
+} while(0)
+
+	/* special cases where incoming command needs to be adjusted */
+	do {
+		if (s->is_subsystem >= SUBSYSTEM_INT_SFTP_ERROR) {
+			command = "echo This service allows sftp connections only.";
+			break;
+		}
+
+		/* if scp or sftp - add module path if command is not absolute */
+		if (s->is_subsystem || (command && memcmp(command, "scp", 3) == 0)) {
```

Comment:
not sure command is always lower case. If not, use _memicmp

Reply:
lets keep it this way for now. OpenSSH client always sends lower case.